lib/autokey/scripting/engine.py

```diff
@@ -47,19 +49,91 @@ def get_folder(self, title: str):
         Note that if more than one folder has the same title, only the first match will be
         returned.
         """
+        validateType(title, "title", str)
         for folder in self.configManager.allFolders:
             if folder.title == title:
                 return folder
         return None
 
+    def create_folder(self, title: str, parent_folder=None, temporary=False):
+        """
+        Create and return a new folder.
+
+        Usage: C{engine.create_folder("new folder"), parent_folder=folder, temporary=True}
+
+        Descriptions for the optional arguments:
+
+        @param parentFolder: Folder to make this folder a subfolder of. If
+        passed as a folder, it will be that folder within auotkey.
+        If passed as pathlib.Path, it will be created or added at that path.
+        Paths expand ~ to $HOME.
+        @param temporary: Folders created with temporary=True are
+                                    not persisted.
+                                    Used for single-source rc-style scripts.
+                                    Cannot be used if parent_folder is a Path.
+
+        If a folder of that name already exists, this will return it unchanged.
+        If the folder wasn't already added to autokey, it will be.
+        The 'temporary' property is not touched to avoid deleting an existing
+        folder.
+        Note that if more than one folder has the same title, only the first match will be
+        returned.
+        """
+        validateType(title, "title", str)
+        validateType(parent_folder, "parent_folder",
+                [model.Folder, pathlib.Path])
+        validateType(temporary, "temporary", bool)
+        # XXX Doesn't check if a folder already exists at this path in autokey.
```

Comment:
I thought a bit about this existence check (This is not a request to actually implement this, I’m just brainstorming a bit):

A simple approach (not 100% foolproof) is checking if the path is a descendant of any existing top level path. To also catch symbolic links, both the given path and all of the autokey top level folder paths have to be resolved and made absolute. If it is, add all directory levels in between in case they are missing, then add the requested folder.

But the above still does not catch hard links (It’s not possible on Linux to create hard directory links) and `bind` mounts. The latter either requires parsing `/etc/mtab` (which may contain wrong data…) or using another approach:

Take the absolute, given path and iterate to the root. At each step, check:
- Does the current path contain an AutoKey json file. If so, stop the iteration.
- Then check if the found path is actually registered in AutoKey (you may have found a backup or smth). See above. This step still has issues with bind mounts, unless you start being tricky: Iterate further up until you end finding AK json files (so you found the top level), saving the Folder names on the way in a list. Use that list to look up the path inside the AutoKey data structure. Beware the "duplicate names permittet" issue. To solve that, also examine the numerical suffixes.
- If the given path is multiple levels below the found path, add all in-between as descendants to AutoKey, Then add the given path.